src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -729,8 +753,8 @@ static ssize_t xgq_transfer_data(struct xocl_xgq_vmr *xgq, const void *buf,
 
 	/* wait for command completion */
 	if (wait_for_completion_killable(&cmd->xgq_cmd_complete)) {
-		XGQ_ERR(xgq, "submit cmd killed");
-		goto done;
+		XGQ_ERR(xgq, "submitted cmd killed");
+		xgq_submitted_cmd_remove(xgq, cmd);
```

Comment:
why 'goto done' is not needed here? Maybe you can have a unified code to call xgq_submitted_cmd_remove() once instead of calling it on every error branch?

Reply:
how to call it once? you mean a callback API?

We indeed need to improve the code structure but maybe not now because it will introduce too many code changes.

The goto is removed, so that the xgq_cmd_rcode can be checked after xgq_submitted_cmd_remove. It should return -EIO. If we goto done, we return 0 which is incorrect.

Reply:
Ah. Okay, they are in different functions. I thought they are in the same function. :)